src/mlpack/core/optimizers/sdp/primal_dual_impl.hpp

```diff
@@ -0,0 +1,386 @@
+
+  arma::vec rhs = rp;
+  const size_t numConstraints = Asparse.n_rows + Adense.n_rows;
+  if (Asparse.n_rows)
+    rhs(arma::span(0, Asparse.n_rows - 1)) += Asparse * Einv_Frd_rc;
+  if (Adense.n_rows)
+    rhs(arma::span(Asparse.n_rows, numConstraints - 1)) += Adense * Einv_Frd_rc;
+
+  // TODO(stephentu): use a more efficient method (e.g. LU decomposition)
+  if (!arma::solve(dy, M, rhs))
+    Log::Fatal << "PrimalDualSolver::SolveKKTSystem(): Could not solve KKT system" << std::endl;
+
+  if (Asparse.n_rows)
+    dysparse = dy(arma::span(0, Asparse.n_rows - 1));
+  if (Adense.n_rows)
+    dydense = dy(arma::span(Asparse.n_rows, numConstraints - 1));
+
+  math::Smat(F * (rd - Asparse.t() * dysparse - Adense.t() * dydense) - rc, Frd_ATdy_rc_Mat);
+  SolveLyapunov(Einv_Frd_ATdy_rc_Mat, Z, 2. * Frd_ATdy_rc_Mat);
+  math::Svec(Einv_Frd_ATdy_rc_Mat, Einv_Frd_ATdy_rc);
+  dsx = -Einv_Frd_ATdy_rc;
+  dsz = rd - Asparse.t() * dysparse - Adense.t() * dydense;
+}
+
+namespace private_ {
+
+// TODO(stephentu): should we move this somewhere more general
+template <typename T> struct vectype { };
+template <typename eT> struct vectype<arma::Mat<eT>> { typedef arma::Col<eT> type; };
+template <typename eT> struct vectype<arma::SpMat<eT>> { typedef arma::SpCol<eT> type; };
+
+} // namespace private_
+
+template <typename SDPType>
+double
+PrimalDualSolver<SDPType>::Optimize(arma::mat& X,
+                                    arma::vec& ysparse,
+                                    arma::vec& ydense,
+                                    arma::mat& Z)
+{
+  // TODO(stephentu): We need a method which deals with the case when the Ais
+  // are not linearly independent.
+
+  const size_t n = sdp.N();
+  const size_t n2bar = sdp.N2bar();
+
+  arma::sp_mat Asparse(sdp.NumSparseConstraints(), n2bar);
+  arma::sp_vec Aisparse;
+
+  for (size_t i = 0; i < sdp.NumSparseConstraints(); i++)
+  {
+    math::Svec(sdp.SparseA()[i], Aisparse);
+    Asparse.row(i) = Aisparse.t();
+  }
+
+  arma::mat Adense(sdp.NumDenseConstraints(), n2bar);
+  arma::vec Aidense;
+  for (size_t i = 0; i < sdp.NumDenseConstraints(); i++)
+  {
+    math::Svec(sdp.DenseA()[i], Aidense);
+    Adense.row(i) = Aidense.t();
+  }
+
+  typename private_::vectype<typename SDPType::objective_matrix_type>::type sc;
+  math::Svec(sdp.C(), sc);
+
+  X = initialX;
+  ysparse = initialYsparse;
+  ydense = initialYdense;
+  Z = initialZ;
+
+  arma::vec sx, sz, dysparse, dydense, dsx, dsz;
+  arma::mat dX, dZ;
+
+  math::Svec(X, sx);
+  math::Svec(Z, sz);
+
+  arma::vec rp, rd, rc, gk;
+
+  arma::mat Rc, F, Einv_F_AsparseT, Einv_F_AdenseT, Gk,
+            M, DualCheck;
+
+  rp.set_size(sdp.NumConstraints());
+
+  Einv_F_AsparseT.set_size(n2bar, sdp.NumSparseConstraints());
+  Einv_F_AdenseT.set_size(n2bar, sdp.NumDenseConstraints());
+  M.set_size(sdp.NumConstraints(), sdp.NumConstraints());
+
+  double primalObj = 0., alpha, beta;
+  for (size_t iteration = 0; iteration != maxIterations; iteration++)
```

Comment:
`iteration` should probably initialize to 1 if you want to allow unlimited iterations with `maxIterations = 0`.